vim-patch: 'completefuzzycollect' option

zeertzjq opened the pull request:
#### vim-patch:9.1.0598: fuzzy completion does not work with default completion

Problem:  fuzzy completion does not work with default completion
Solution: Make it work (glepnir)

closes: vim/vim#15193

https://github.com/vim/vim/commit/8159fb18a92e9a9f5e35201bd92bf651f4d5835c

Cherry-pick insexpand.c changes from patch 9.1.0608.

N/A patch:
vim-patch:9.1.0632: MS-Windows: Compiler Warnings

Co-authored-by: glepnir <glephunter@gmail.com>


#### vim-patch:9.1.0605: internal error with fuzzy completion

Problem:  internal error with fuzzy completion
          (techntools)
Solution: only fuzzy complete the pattern after directory separator
          (glepnir)

closes: vim/vim#15291

https://github.com/vim/vim/commit/0be03e14b9a2899f5e96720e3b21935bd9d2d34e

Co-authored-by: glepnir <glephunter@gmail.com>


#### vim-patch:9.1.0631: wrong completion list displayed with non-existing dir + fuzzy completion

Problem:  wrong completion list displayed with non-existing dir + fuzzy
          completion (kawarimidoll)
Solution: clear list of matches, if leader did not use fuzzy match
          (glepnir)

closes: vim/vim#15365

https://github.com/vim/vim/commit/6b6280c4a270547f84f01c0e0d9be1b7d6bb9e20

Co-authored-by: glepnir <glephunter@gmail.com>


#### vim-patch:9.1.0634: Ctrl-P not working by default

Problem:  Ctrl-P not working by default
          (Jesse Pavel, after v9.1.0598)
Solution: Revert part of v9.1.0598 and set cur_match_pos
          correctly according to compl_dir_forward()

closes: vim/vim#15379

https://github.com/vim/vim/commit/13032a49b7d2a45e7c774cf23ee8f58f56b03781

Co-authored-by: Christian Brabandt <cb@256bit.org>


#### vim-patch:9.1.0654: completion does not respect completeslash with fuzzy

Problem:  completion does not respect completeslash with fuzzy
          (egesip)
Solution: Change path separator on Windows, depending on 'completeslash'
          option value (glepnir)

closes: vim/vim#15418

https://github.com/vim/vim/commit/b9de1a057f9a0b6de6f64a9c1b2078c7069cdd7d

Co-authored-by: glepnir <glephunter@gmail.com>


#### vim-patch:9.1.0705: Sorting of fuzzy filename completion is not stable

Problem:  Sorting of fuzzy filename completion is not stable
Solution: Compare indexes when scores are equal.  Fix some typos.
          (zeertzjq)

closes: vim/vim#15593

https://github.com/vim/vim/commit/58d705238c0794ee3baa4173831ab157e709a48a


#### vim-patch:9.1.0733: keyword completion does not work with fuzzy

Problem:  keyword completion does not work with fuzzy
          (egesip)
Solution: handle ctrl_x_mode_normal() specifically
          (glepnir)

closes: vim/vim#15424

https://github.com/vim/vim/commit/7cfe693f9bfa74690867e4d96c25f2205d0d13e4

Co-authored-by: glepnir <glephunter@gmail.com>


#### vim-patch:9.1.1006: PmenuMatch completion highlight can be combined

Problem:  PmenuMatch completion highlight can be combined
Solution: Combine highlight groups PmenuMatch with Pmenu and
          PmenuMatchSel with PmenuSel (glepnir)

closes: vim/vim#16408

https://github.com/vim/vim/commit/9eff3ee81839d67999491d293879ada134df2d3e

Co-authored-by: glepnir <glephunter@gmail.com>


#### vim-patch:9.1.1008: tests: test for patch 9.1.1006 doesn't fail without the patch

Problem:  tests: test for patch 9.1.1006 doesn't fail without the patch
          (after v9.1.1006)
Solution: Add ctermbg=NONE to the highlight groups (zeertzjq).

closes: vim/vim#16425

https://github.com/vim/vim/commit/faf250c9e4f6f9fa5db38d9373735e77ba438a90


#### vim-patch:9.1.1131: potential out-of-memory issue in search.c

Problem:  potential out-of-memory issue in search.c
Solution: improve situation and refactor search.c slightly
          (John Marriott)

- In function update_search_stat():
  add a check for a theoretical null pointer reference, set and remember
  the length of lastpat, remove the three calls to STRLEN() and use the
  various string's associated lengths instead, add a check for an
  out-of-memory condition.

- In function search_for_fuzz_match():
  remove a call to strnsave() and thus avoid having to add a check for
  an out-of-memory condition, also replace the call to STRLEN() by
  ml_get_buf_len().

closes: vim/vim#16689

https://github.com/vim/vim/commit/b79fa3d9c8a08f15267797511d779e33bd33e68e

Co-authored-by: John Marriott <basilisk@internode.on.net>


#### vim-patch:9.1.1178: not possible to generate completion candidates using fuzzy matching

Problem:  not possible to generate completion candidates using fuzzy
          matching
Solution: add the 'completefuzzycollect' option for (some) ins-completion
          modes (glepnir)

closes: vim/vim#16032

https://github.com/vim/vim/commit/f31cfa29bf72b0cdf6fa1b60346ea4e187bcafd1

Co-authored-by: glepnir <glephunter@gmail.com>


#### vim-patch:9.1.1181: Unnecessary STRLEN() calls in insexpand.c

Problem:  Unnecessary STRLEN() calls in insexpand.c (after 9.1.1178).
Solution: Use the already available length (zeertzjq).

closes: vim/vim#16814

https://github.com/vim/vim/commit/4422de6316b544c282e6c74afd3df3ee3a4b1cfd


#### vim-patch:9.1.1182: No cmdline completion for 'completefuzzycollect'

Problem:  No cmdline completion for the 'completefuzzycollect' option
          (after v9.1.1178)
Solution: Add cmdline completion for the 'completefuzzycollect' option,
          improve its description in optwin.vim (zeertzjq).

closes: vim/vim#16813

https://github.com/vim/vim/commit/53d59ecc1d93ce3a3f6d0182479d825852018ceb

No code change is needed in Nvim as Nvim uses expand_set_str_generic()
by default.


#### vim-patch:9.1.1185: endless loop with completefuzzycollect and no match found

Problem:  endless loop with completefuzzycollect and no match found
Solution: move pointer to line end and break loop

closes: vim/vim#16820

https://github.com/vim/vim/commit/dd42b05f8a37df03a9b77a16a47c08ab33af2b1f

Co-authored-by: glepnir <glephunter@gmail.com>


#### vim-patch:9.1.1197: process_next_cpt_value() uses wrong condition

Problem:  process_next_cpt_value() uses wrong condition
Solution: use cfc_has_mode() instead and remove redundant else if branch
          (glepnir)

closes: vim/vim#16833

https://github.com/vim/vim/commit/53b14578e03f93a53fd6eb21c00caf96484742ed

Co-authored-by: glepnir <glephunter@gmail.com>


#### vim-patch:9.1.1201: 'completefuzzycollect' does not handle dictionary correctly

Problem:  'completefuzzycollect' does not handle dictionary correctly
Solution: check for ctrl_x_mode_dictionary (glepnir)

closes: vim/vim#16867

https://github.com/vim/vim/commit/587601671cd06ddc4d78f907d98578cdab96003f

Cherry-pick a documentation fix from later.

Co-authored-by: glepnir <glephunter@gmail.com>


#### vim-patch:9.1.1228: completion: current position column wrong after got a match

Problem:  The current_pos.col was incorrectly updated to the length of
          the matching text. This will cause the next search to start
          from the wrong position.
Solution: current_pos has already been updated in search_str_in_line and
          does not need to be changed (glepnir)

closes: vim/vim#16941

https://github.com/vim/vim/commit/5753084042e17d794627d77e4300def031ce5498

Co-authored-by: glepnir <glephunter@gmail.com>

zeertzjq committed:
vim-patch:9.1.0598: fuzzy completion does not work with default completion

Problem:  fuzzy completion does not work with default completion
Solution: Make it work (glepnir)

closes: vim/vim#15193

https://github.com/vim/vim/commit/8159fb18a92e9a9f5e35201bd92bf651f4d5835c

Cherry-pick insexpand.c changes from patch 9.1.0608.

N/A patch:
vim-patch:9.1.0632: MS-Windows: Compiler Warnings

Co-authored-by: glepnir <glephunter@gmail.com>

zeertzjq committed:
vim-patch:9.1.0605: internal error with fuzzy completion

Problem:  internal error with fuzzy completion
          (techntools)
Solution: only fuzzy complete the pattern after directory separator
          (glepnir)

fixes: vim/vim#15287
closes: vim/vim#15291

https://github.com/vim/vim/commit/0be03e14b9a2899f5e96720e3b21935bd9d2d34e

Co-authored-by: glepnir <glephunter@gmail.com>

zeertzjq committed:
vim-patch:9.1.0631: wrong completion list displayed with non-existing dir + fuzzy completion

Problem:  wrong completion list displayed with non-existing dir + fuzzy
          completion (kawarimidoll)
Solution: clear list of matches, if leader did not use fuzzy match
          (glepnir)

fixes: vim/vim#15357
closes: vim/vim#15365

https://github.com/vim/vim/commit/6b6280c4a270547f84f01c0e0d9be1b7d6bb9e20

Co-authored-by: glepnir <glephunter@gmail.com>

zeertzjq committed:
vim-patch:9.1.0634: Ctrl-P not working by default

Problem:  Ctrl-P not working by default
          (Jesse Pavel, after v9.1.0598)
Solution: Revert part of v9.1.0598 and set cur_match_pos
          correctly according to compl_dir_forward()

fixes: vim/vim#15370
closes: vim/vim#15379

https://github.com/vim/vim/commit/13032a49b7d2a45e7c774cf23ee8f58f56b03781

Co-authored-by: Christian Brabandt <cb@256bit.org>

zeertzjq committed:
vim-patch:9.1.0654: completion does not respect completeslash with fuzzy

Problem:  completion does not respect completeslash with fuzzy
          (egesip)
Solution: Change path separator on Windows, depending on 'completeslash'
          option value (glepnir)

fixes: vim/vim#15392
closes: vim/vim#15418

https://github.com/vim/vim/commit/b9de1a057f9a0b6de6f64a9c1b2078c7069cdd7d

Co-authored-by: glepnir <glephunter@gmail.com>

glepnir commented:
could you mark as draft..I am not port it  because there are still some issues to discuss..

zeertzjq committed:
vim-patch:9.1.0705: Sorting of fuzzy filename completion is not stable

Problem:  Sorting of fuzzy filename completion is not stable
Solution: Compare indexes when scores are equal.  Fix some typos.
          (zeertzjq)

closes: vim/vim#15593

https://github.com/vim/vim/commit/58d705238c0794ee3baa4173831ab157e709a48a

zeertzjq committed:
vim-patch:26e4b00: runtime(doc): Revert outdated comment in completeopt's fuzzy documentation

Originally, `:set completeopt+=fuzzy` did not affect how the candidate
list is collected in default keyword completion. A comment was added to
documentation as part of vim/vim#14912 to clarify it. vim/vim#15193 later changed the
fuzzy behavior to now change the candidate collection behavior as well
so the clarification in docs is now wrong. Remove them here.

closes: vim/vim#15656

https://github.com/vim/vim/commit/26e4b000025ea0e25ea7877314d9095737431bae

Co-authored-by: Yee Cheng Chin <ychin.git@gmail.com>

justinmk commented on runtime/doc/options.txt:
```diff
@@ -1566,10 +1566,7 @@ A jump table for the options with a short description can be found at |Q_op|.
-		    difference how completion candidates are reduced from the
-		    list of alternatives, but not how the candidates are
-		    collected (using different completion types).
+		    if the exact sequence is not typed.
```
Nice

zeertzjq committed:
vim-patch:9.1.0733: keyword completion does not work with fuzzy

Problem:  keyword completion does not work with fuzzy
          (egesip)
Solution: handle ctrl_x_mode_normal() specifically
          (glepnir)

fixes: vim/vim#15412
closes: vim/vim#15424

https://github.com/vim/vim/commit/7cfe693f9bfa74690867e4d96c25f2205d0d13e4

Co-authored-by: glepnir <glephunter@gmail.com>

zeertzjq committed:
vim-patch:9.1.1006: PmenuMatch completion highlight can be combined

Problem:  PmenuMatch completion highlight can be combined
Solution: Combine highlight groups PmenuMatch with Pmenu and
          PmenuMatchSel with PmenuSel (glepnir)

fixes: vim/vim#15563
closes: vim/vim#16408

https://github.com/vim/vim/commit/9eff3ee81839d67999491d293879ada134df2d3e

Co-authored-by: glepnir <glephunter@gmail.com>

zeertzjq committed:
vim-patch:9.1.1008: tests: test for patch 9.1.1006 doesn't fail without the patch

Problem:  tests: test for patch 9.1.1006 doesn't fail without the patch
          (after v9.1.1006)
Solution: Add ctermbg=NONE to the highlight groups (zeertzjq).

closes: vim/vim#16425

https://github.com/vim/vim/commit/faf250c9e4f6f9fa5db38d9373735e77ba438a90

zeertzjq committed:
vim-patch:9.1.1131: potential out-of-memory issue in search.c

Problem:  potential out-of-memory issue in search.c
Solution: improve situation and refactor search.c slightly
          (John Marriott)

- In function update_search_stat():
  add a check for a theoretical null pointer reference, set and remember
  the length of lastpat, remove the three calls to STRLEN() and use the
  various string's associated lengths instead, add a check for an
  out-of-memory condition.

- In function search_for_fuzz_match():
  remove a call to strnsave() and thus avoid having to add a check for
  an out-of-memory condition, also replace the call to STRLEN() by
  ml_get_buf_len().

closes: vim/vim#16689

https://github.com/vim/vim/commit/b79fa3d9c8a08f15267797511d779e33bd33e68e

Co-authored-by: John Marriott <basilisk@internode.on.net>

zeertzjq committed:
vim-patch:9.1.1178: not possible to generate completion candidates using fuzzy matching

Problem:  not possible to generate completion candidates using fuzzy
          matching
Solution: add the 'completefuzzycollect' option for (some) ins-completion
          modes (glepnir)

fixes vim/vim#15296
fixes vim/vim#15295
fixes vim/vim#15294
closes: vim/vim#16032

https://github.com/vim/vim/commit/f31cfa29bf72b0cdf6fa1b60346ea4e187bcafd1

Co-authored-by: glepnir <glephunter@gmail.com>

zeertzjq commented on test/old/testdir/test_ins_complete.vim:
```diff
@@ -2890,13 +2870,240 @@ func Test_complete_fuzzy_match()
 
+func Test_complete_fuzzy_collect()
+  new
+  redraw  " need this to prevent NULL dereference in Nvim
```
`curwin->w_grid` is not allocated until a redraw, so without this `:redraw` there is a NULL dereference in `pum_display()`. Not sure if that should be fixed, as it doesn't really happen in practice.

zeertzjq commented:
With the `'completefuzzycollect'` option this is now backward-compatible, and `'completefuzzycollect'` is a bit new and currently not particularly useful, so I'll postpone this after the v0.11 release.

glepnir commented:
okay

zeertzjq committed:
vim-patch:9.1.1181: Unnecessary STRLEN() calls in insexpand.c

Problem:  Unnecessary STRLEN() calls in insexpand.c (after 9.1.1178).
Solution: Use the already available length (zeertzjq).

closes: vim/vim#16814

https://github.com/vim/vim/commit/4422de6316b544c282e6c74afd3df3ee3a4b1cfd

zeertzjq committed:
vim-patch:9.1.1182: No cmdline completion for 'completefuzzycollect'

Problem:  No cmdline completion for the 'completefuzzycollect' option
          (after v9.1.1178)
Solution: Add cmdline completion for the 'completefuzzycollect' option,
          improve its description in optwin.vim (zeertzjq).

closes: vim/vim#16813

https://github.com/vim/vim/commit/53d59ecc1d93ce3a3f6d0182479d825852018ceb

No code change is needed in Nvim as Nvim uses expand_set_str_generic()
by default.

zeertzjq committed:
vim-patch:9.1.1185: endless loop with completefuzzycollect and no match found

Problem:  endless loop with completefuzzycollect and no match found
Solution: move pointer to line end and break loop

closes: vim/vim#16820

https://github.com/vim/vim/commit/dd42b05f8a37df03a9b77a16a47c08ab33af2b1f

Co-authored-by: glepnir <glephunter@gmail.com>

zeertzjq committed:
vim-patch:9.1.1197: process_next_cpt_value() uses wrong condition

Problem:  process_next_cpt_value() uses wrong condition
Solution: use cfc_has_mode() instead and remove redundant else if branch
          (glepnir)

closes: vim/vim#16833

https://github.com/vim/vim/commit/53b14578e03f93a53fd6eb21c00caf96484742ed

Co-authored-by: glepnir <glephunter@gmail.com>

zeertzjq committed:
vim-patch:1dc731a: runtime(doc): make :h 'completefuzzycollect' a bit clearer

- Fix grammar
- Use "matches" instead of "items" ("completion candidates" is used in
  some other places, but it's a bit verbose)
- "When set" is a bit vague, instead use "For specified modes"

closes: vim/vim#16871

https://github.com/vim/vim/commit/1dc731a49ff5d06ead4b506afa04288a5baafc1a

zeertzjq committed:
vim-patch:9.1.1201: 'completefuzzycollect' does not handle dictionary correctly

Problem:  'completefuzzycollect' does not handle dictionary correctly
Solution: check for ctrl_x_mode_dictionary (glepnir)

closes: vim/vim#16867

https://github.com/vim/vim/commit/587601671cd06ddc4d78f907d98578cdab96003f

Cherry-pick a documentation fix from later.

Co-authored-by: glepnir <glephunter@gmail.com>

zeertzjq committed:
vim-patch:9.1.1228: completion: current position column wrong after got a match

Problem:  The current_pos.col was incorrectly updated to the length of
          the matching text. This will cause the next search to start
          from the wrong position.
Solution: current_pos has already been updated in search_str_in_line and
          does not need to be changed (glepnir)

closes: vim/vim#16941

https://github.com/vim/vim/commit/5753084042e17d794627d77e4300def031ce5498

Co-authored-by: glepnir <glephunter@gmail.com>
